### PR TITLE
NavLink, NavButton and RouteRefs enforcing

### PR DESCRIPTION
**Pull request by shmidt-i**
## Context
1. At the moment, a lot of places are using the Link/Button components directly from the `@material-ui/core`, which effectively bypasses react-router and leads to app reinstantiation after each navigation. 
2. Also, new RouteRef API is missing in the documentation and templates 

## This PR includes:
1. Link & Button components - thin layer of abstraction on top of material-ui Link and Button components correspondingly, sets default for the `component` prop to be `Link` from the `react-router`. This automatically enables the resulting component to accept props for either of the underlying components.
2. More info about the new RouteRef API in the docs and changes in the template to create routes the new way.
3. Removed the `registerRedirect` API for now

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Few comments about redirects. As mentioned I think we can fully remove them for now and then reintroduce after rethinking how they need to work.

I'm also not 100% sure about the `NavButton` and `NavLink` naming. Internally we have wrappers for both of those that handle more things than just navigation with react-router, so I'm thinking the name itself isn't really forwards-compatible. Might be better to just call them `Button` and `Link`. Either way we can add lint rules that recommend usage of those components instead of the MUI ones.

File: docs/reference/createPlugin-router.md [hunk 1]
```diff
@@ -1,21 +1,48 @@
...
-    options?: RouteOptions,
-  ): void;
-};
+addRoute(
+  target: RouteRef,
+  Component: ComponentType<any>,
+  options?: RouteOptions,
+): void;
+
+addRedirect(from: RouteRef, to: RouteRef, options?: RouteOptions): void;
```
[the remark this answers is not in the document]
Comment (shmidt-i) [reply docs/reference/createPlugin-router.md:13]: Okay, will remove 👌

File: docs/reference/createPlugin-router.md [hunk 2]
```diff
@@ -1,21 +1,48 @@
...
+
+## RouteRef
+
+Both `addRoute` and `addRedirect` methods are using mutable RouteRefs, which can be created as following:
+
+```ts
+import { createRouteRef } from '@backstage/core';
+
+const myPluginRouteRef = createRouteRef({
+  icon: () => null, // You can set an icon for your route
```
[the remark this answers is not in the document]
Comment (shmidt-i) [reply docs/reference/createPlugin-router.md:42]: Yes, I like it!

**Comment by shmidt-i** [comment]
Ready for another look @Rugvip ;)

File: packages/dev-utils/src/devApp/render.tsx [hunk 3]
```diff
@@ -148,7 +149,7 @@ class DevAppBuilder {
                 key={target.path}
                 to={target.path}
                 text={target.title}
-                icon={target.icon}
+                icon={target.icon ?? SentimentDissatisfiedIcon}
```
Comment (Rugvip) [inline packages/dev-utils/src/devApp/render.tsx:152]: 👌

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**